tamr_client/project.py

```diff
@@ -198,3 +208,28 @@ def get_all(
         project = _from_json(project_url, project_json)
         projects.append(project)
     return tuple(projects)
+
+
+def attributes(session: Session, project: Project) -> Tuple[Attribute, ...]:
+    """Get all attributes from a project
+
+    Args:
+        project: Project containing the desired attributes
+
+    Returns:
+        The attributes for the specified project
+
+    Raises:
+        requests.HTTPError: If an HTTP error is encountered.
+    """
+    attrs_url = replace(project.url, path=project.url.path + "/attributes")
+    r = session.get(str(attrs_url))
+    attrs_json = response.successful(r).json()
+
+    attrs = []
+    for attr_json in attrs_json:
+        id = attr_json["name"]
+        attr_url = replace(attrs_url, path=attrs_url.path + f"/{id}")
+        attr = _attribute_from_json(attr_url, attr_json)
+        attrs.append(attr)
+    return tuple(attrs)
```

Comment:
To avoid mutation logic and type conversion, you could define a helper function and use a tuple comprehension or `map` call:

```suggestion
    def attr_from_json(attr_json):
        id = attr_json["name"]
        attr_url = replace(attrs_url, path=attrs_url.path + f"/{id}")
        return _attribute_from_json(attr_url, attr_json)
    return tuple(map(attr_from_json, attrs_json))
```

Reply:
The `project.attributes` and `dataset.attributes` functions only differ in their typehints and docstrings.  They are logically identical.  Is there a better practice than having this repeated code, or is the clarity of having them defined separately more valuable?

Reply:
I think its a question of whether the repeated logic has an intuitive tag/name. If yes, make a shared utility function with that name. If no and the shared logic is small, its not too bad to keep some minor duplication.

Reply:
This case does make sense to have a shared utility function `attribute._get_all_from_parent`, so I've done that.